In [1]:
!pip install fiona
!pip install geopandas
!pip install folium
!pip install sodapy -q
!pip install adjustText
!pip install plotly --upgrade
!pip install contextily
!pip install mapclassify
!pip install chart-studio
!sudo apt install libspatialindex-dev
!pip install rtree
!pip install osmnx
!pip install requests

!pip install matplotlib==3.1.3
!pip install js2py
!pip install selenium

     |████████████████████████████████| 16.7 MB 254 kB/s 
     |████████████████████████████████| 1.0 MB 6.7 MB/s 
     |████████████████████████████████| 6.3 MB 32.0 MB/s 
  Created wheel for adjustText: filename=adjustText-0.7.3-py3-none-any.whl size=7097 sha256=9b72b0c5d01f41cfda53113636b8eb2ac20f4ce40d060200be2da93572d5dbea
  Stored in directory: /root/.cache/pip/wheels/2f/98/32/afbf902d8f040fadfdf0a44357e4ab750afe165d873bf5893d
Successfully built adjustText
     |████████████████████████████████| 28.8 MB 1.4 MB/s 
  Attempting uninstall: plotly
    Found existing installation: plotly 5.5.0
    Uninstalling plotly-5.5.0:
      Successfully uninstalled plotly-5.5.0
     |████████████████████████████████| 19.3 MB 445 kB/s 
     |████████████████████████████████| 64 kB 1.8 MB/s 
  Created wheel for retrying: filename=retrying-1.3.3-py3-none-any.whl size=11447 sha256=1871e963a265211d27a521cf70d668cee25fb3f9684d2da6d1e9c4142028ff23
  Stored in directory: /root/.cache/pip/wheels/f9/8d/8d

     |████████████████████████████████| 13.1 MB 5.0 MB/s 
  Attempting uninstall: matplotlib
    Found existing installation: matplotlib 3.5.1
    Uninstalling matplotlib-3.5.1:
      Successfully uninstalled matplotlib-3.5.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
osmnx 1.1.2 requires matplotlib>=3.4, but you have matplotlib 3.1.3 which is incompatible.
albumentations 0.1.12 requires imgaug<0.2.7,>=0.2.5, but you have imgaug 0.2.9 which is incompatible.


     |████████████████████████████████| 1.0 MB 5.0 MB/s 
  Created wheel for pyjsparser: filename=pyjsparser-2.7.1-py3-none-any.whl size=26000 sha256=4e5c9e477d9855a215c6e4fa6350eed42253b46189b60b8c88af1fb84c1e42ad
  Stored in directory: /root/.cache/pip/wheels/2a/80/ac/dcd2bdbd03dd2b7b7e2bf3e5afbda6a1ab7935bbce314969da
Successfully built pyjsparser
     |████████████████████████████████| 968 kB 5.1 MB/s 
     |████████████████████████████████| 359 kB 47.7 MB/s 
     |████████████████████████████████| 138 kB 54.3 MB/s 
     |████████████████████████████████| 55 kB 3.4 MB/s 
     |████████████████████████████████| 4.0 MB 44.2 MB/s 
     |████████████████████████████████| 58 kB 5.4 MB/s 
  Attempting uninstall: urllib3
    Found existing installation: urllib3 1.24.3
    Uninstalling urllib3-1.24.3:
      Successfully uninstalled urllib3-1.24.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the fol

In [2]:
# To ignore unimporant system warnings
import warnings
warnings.filterwarnings("ignore")

# We will use Pandas, Numpy, and Matplotlib which is a package for visualization with Python
import pandas as pd
import numpy as np

# Load a required package 
# This is a library for accessing and parsing data through URLs
from urllib.parse import urlencode
import urllib.request, json 
from bs4 import BeautifulSoup # for web scraping
import requests
import fiona
import folium
import plotly.express as px
import geopandas as gpd
import adjustText as aT
import seaborn as sns # visualization styling package
import plotly.graph_objects as go
import urllib, json
import chart_studio.plotly as py
import chart_studio.tools as tls
import js2py
import os
import osmnx as ox
from pandas.io.json import json_normalize
import seaborn as sns
from sodapy import Socrata
from IPython.display import Image, display
import copy
from shapely.geometry import Polygon
import time
import ast
from shapely.geometry import shape
import contextily as ctx
import re
import pickle
from urllib.request import urlopen

# A magic functin that renders the figure in a notebook 
%matplotlib inline 

import matplotlib
matplotlib.use('Agg')

from matplotlib import pyplot as plt
from shapely.geometry import Point
from shapely.geometry import MultiPoint
from shapely import wkt

from google.colab import files

%matplotlib inline

!wget -cq https://raw.githubusercontent.com/kirthi-b/Urban-Mobility-Index/main/Script/jscode/googlepolydecoder.js
!wget -cq https://raw.githubusercontent.com/kirthi-b/Urban-Mobility-Index/main/Data/Training/Training_Class.csv
!wget -cq https://raw.githubusercontent.com/kirthi-b/Urban-Mobility-Index/main/Data/Testing/Test_Class.csv

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [4]:
train_class = pd.read_csv('/content/Training_Class.csv')
test_class = pd.read_csv('/content/Test_Class.csv')
#append Train and Test 
cities = train_class.append(test_class)

In [5]:
def class_map(value):

  dfclass = {
    91: "91-100",
    81: "81-90",
    71: "71-80",
    61: "61-70",
    51: "51-60",
    41: "41-50",
    31: "31-40",
    21: "21-30",
    11: "11-20",
    0: "0-10",}

  for key, score in dfclass.items():
     if value >= key:
       return score

In [6]:
cities = cities[['Walk Score','Bus_perSqKm',	'Bus_perThouCap',	'Inter_perSqKm',	'Inter_perThouCap']]

In [7]:
X, y = cities.iloc[:,1:], cities.iloc[:, 0]
# Split train and test dataset

X_train, X_test, y_train, y_test = train_test_split(X,y, train_size = 0.75, random_state=0)

# Check the number of samples
print ("Train sample (X):", len(X_train))
print ("Train sample (y):", len(y_train))
print ("Test sample (X):", len(X_test))
print ("Test sample (y):", len(y_test))

Train sample (X): 580
Train sample (y): 580
Test sample (X): 194
Test sample (y): 194


In [15]:
lr = LinearRegression().fit(X_train, y_train)
lr

LinearRegression()

In [9]:
from sklearn import metrics
y_pred = lr.predict(X_test)

d = y_test - y_pred
r2_f = 1-(sum(d**2)/sum((y_test - np.mean(y_test))**2))

print('Mean Walk Score:', cities['Walk Score'].mean())
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))
print("R-Squared:", r2_f)

Mean Walk Score: 71.0749354005168
Root Mean Squared Error: 17.039380028393133
R-Squared: 0.3816855292507172


In [10]:
print ("Class:",class_map(lr.predict([[10, 20, 500, 500]])))

Class: 61-70


value of root mean squared error is 17.04, which is less than 25% of the mean value of the all walkscores i.e. 71.07. It's not great, but not terrible.

In [11]:
def getneigh(state, city):

  #nodes here is a dataframe containing node info of a city

  # Create an URL object
  city = city.replace(" ", "_")
  url = 'https://www.walkscore.com/'+ state +'/'+ city +'/'

  # Create object page
  page = requests.get(url)
  soup = BeautifulSoup(page.text, 'lxml')

  # Obtain information from tag <table>
  table1 = soup.find('table', id='hoods-list-table')

  # Obtain every title of columns with tag <th>
  headers = []
  for i in table1.find_all('th'):
   title = i.text
   headers.append(title)

  df = pd.DataFrame(columns = headers)
  
  # Create a for loop to fill df
  for j in table1.find_all('tr')[1:]:
   row_data = j.find_all('td')
   row = [i.text for i in row_data]
   length = len(df)      
   df.loc[length] = row

  #Get URL of each neighborhood
  df['URL'] = url + df['Name'].str.replace(" ", "_")

  return df
  #return df['Name'].values.tolist()

In [12]:
def getmapdensity(neigh, city): 
    
  df = neigh.reset_index()

  sf = city + ', usa'
  # Get graph by geocoding
  graph_sf = ox.graph_from_place(sf, network_type="walk")
  crs = 4326

  # Project graph
  graph_sf = ox.projection.project_graph(graph_sf, to_crs=crs)

  # everything to gdfs
  nodes = ox.graph_to_gdfs(graph_sf)[0]

  #START FUNCTIONS

  def getpoly(url):
    #url would be the url of the neighborhood's/city's page
    data = urlopen(url)
    source = data.read().decode('utf-8')

    #regex to search
    regex = r"https:\/\/pp\.walk\.sc\/_\/s\/_p\/hood\/[a-z0-9]+\.js"

    matches = re.finditer(regex, source, re.MULTILINE)
  
    for matchNum, match in enumerate(matches, start=1):
        jslink = match.group()

    #jslink is the javascript page link for the google map overlay
    data = urlopen(jslink)
    jspage = data.read().decode('utf-8')

    regex = r"(?<=window\.activeEntity = reconstituteEntity)\(.*\)"

    matches = re.finditer(regex, jspage, re.MULTILINE)

    for matchNum, match in enumerate(matches, start=1):
     polycode = match.group()

    #get overlay dictionary
    dic = json.loads(polycode[1:-1])

    #get encoded path string from dictionary
    path = dic['path']

    return path

  def polydecode(en):
    eval_res, tempfile = js2py.run_file("googlepolydecoder.js")
    return tempfile.decodePolyline(en).to_list()

  def getpolybackup(url):
    #url would be the url of the neighborhood's/city's page
    data = urlopen(url)
    source = data.read().decode('utf-8')

    #regex to search
    regex = r"https:\/\/pp\.walk\.sc\/_\/s\/_p\/hood\/[a-z0-9]+\.js"

    matches = re.finditer(regex, source, re.MULTILINE)

    for matchNum, match in enumerate(matches, start=1):
        jslink = match.group()

    #jslink is the javascript page link for the google map overlay
    data = urlopen(jslink)
    jspage = data.read().decode('utf-8')

    regex = r"(?<=window\.activeEntity=reconstituteEntity)(.*)(?=\,window\.activeCity)"

    matches = re.finditer(regex, jspage, re.MULTILINE)
    polycode = ""

    for matchNum, match in enumerate(matches, start=1):
     polycode = match.group()

    regex = r"(?<=path\:)(.*)(?=\,population)"

    matches = re.finditer(regex, polycode, re.MULTILINE)
    path = ""

    for matchNum, match in enumerate(matches, start=1):
     path = match.group()

    return path

  def getBounds(txt):
    dictionary = {"{":'(', "}":')', " ":'', "[":'', "]":''}
    transTable = txt.maketrans(dictionary)
    txt = txt.translate(transTable)
    txt = txt.replace("'latitude':",'')
    txt = txt.replace("'longitude':",'')

    lst = list(ast.literal_eval(txt))
    points = MultiPoint(lst)
    b = points.bounds

    return b

  def getPoly(txt):
    dictionary = {"{":'(', "}":')', " ":'', "[":'', "]":''}
    transTable = txt.maketrans(dictionary)
    txt = txt.translate(transTable)
    txt = txt.replace("'latitude':",'')
    txt = txt.replace("'longitude':",'')
    txt = txt.replace('"','')
    txt = txt.replace(",",' ')
    txt = txt.replace(") (",',')  

    return txt

  #END OF FUNCTIONS


  #Empty Polygon Column
  df["Polygon"] = ""
  df["EnPath"] = ""


  try:
    getpoly(df['URL'][0])
  except:
    a = str(getpolybackup(df['URL'][0]))
  else:
    a = str(getpoly(df['URL'][0]))
  df['EnPath'][0] = a
  df['Polygon'][0] = polydecode(a)
  df["WKT"] = ""
  df["Count"] = ""
  df["WKTPoly"] = ""  
  df["WKTPoly"][0] = copy.deepcopy(df["Polygon"][0])

  for j in range(len(df['Polygon'][0])):
    df['WKTPoly'][0][j] = str(dict(reversed(list(df['WKTPoly'][0][j].items()))))
  if df['WKTPoly'][0][0] != df['WKTPoly'][0][-1]:
    df['WKTPoly'][0].append(df['WKTPoly'][0][0])
  else:
    j = j + 1

  #format WKT column as per geopandas
  url = "https://z.overpass-api.de/api/interpreter"

  p = df['WKTPoly'][0]
  df['WKT'][0] = 'POLYGON (' + getPoly(str(p)) + ')'
  #get bus stop nodes for polygon
  query = f"""
    [out:json][timeout:25];
    node["public_transport"="platform"]["bus"="yes"]
    {getBounds(str(df['Polygon'][0]))};
    out skel;
  """
  r = requests.get(url, params={'data': query})
  #handle errors
  try:
    json_normalize(r.json()['elements'])[['id']]
  except:
    df['Count'][0] = 0
  else:
    #Points Geopandas GeoDataFrame   
    bus = json_normalize(r.json()['elements'])[['id']]  # create a DataFrame from the data
    df['Count'][0] = int(bus.shape[0])

  #drop columns we don't need anymore
  df['geometry'] = df.WKT.apply(wkt.loads)
  df.drop(['Bike Score', 'URL','WKT','Polygon' ,'WKTPoly','EnPath'], axis=1, inplace=True) 

  # Polygon Geopandas GeoDataFrame
  gdf = gpd.GeoDataFrame(df, geometry='geometry', crs = {'init': 'epsg:4326'}) #wgs84 projection - terrible for area calculations
  gdf= gdf.to_crs({'init': 'epsg:2163'}) #equal area projection 
  gdf["area"] = gdf['geometry'].area/ 10**6 #area in sq. km!

  # Calculate Bus Density Fields
  gdf["Bus_perSqKm"] = gdf["Count"]/gdf["area"] #per square kilometer
  gdf['Population'] = gdf['Population'].str.replace(',', '')
  gdf['Population'] = gdf['Population'].astype(int)
  gdf["Bus_perThouCap"] = gdf["Count"]*1000/gdf["Population"] #per thousand capita
  gdf = gdf[['Name','Walk Score','Population','area','geometry','Bus_perSqKm','Bus_perThouCap']]

  #Calculate Intersection Density Fields
  gdf_nodes = nodes.to_crs('epsg:2163')
  gdf_join = gdf.sjoin(gdf_nodes, how="left")
  gdf["Count"] = gdf_join.shape[0]
  gdf = gdf.rename(columns={"perSqKm":"Bus_perSqKm", "perThouCap": "Bus_perThouCap"})
  gdf["Inter_perSqKm"] = gdf["Count"]/gdf["area"] #per square kilometer
  gdf["Inter_perThouCap"] = gdf["Count"]*1000/gdf["Population"] #per thousand capita
  gdf = gdf.drop('Count', 1)
  return gdf

In [13]:
STATE = input("Enter state code: \n")
CITY = input("Enter city: \n")
df = getneigh(STATE, CITY)
print(df[['Name']])
neigh_num = input("Enter index number of neighborhood: \n")

ndf = getmapdensity(df.iloc[[neigh_num]], CITY)
bus_dist = int(ndf['Bus_perSqKm'][0])
bus_pop = int(ndf['Bus_perThouCap'][0])
inter_dist = int(ndf['Bus_perSqKm'][0])
inter_pop= int(ndf['Inter_perThouCap'][0])

print ("Class:",class_map(lr.predict([[bus_dist, bus_pop, inter_dist, inter_pop]])))
print("Approximately 40% of the variability observed in the Walk Scores is explained by the bus stop and intersection densities.")

KeyboardInterrupt: ignored

In [ ]:
STATE = input("Enter state code: \n")
CITY = input("Enter city: \n")
df = getneigh(STATE, CITY)
print(df[['Name']])
neigh_num = input("Enter index number of neighborhood: \n")

ndf = getmapdensity(df.iloc[[neigh_num]], CITY)
bus_dist = int(ndf['Bus_perSqKm'][0])
bus_pop = int(ndf['Bus_perThouCap'][0])
inter_dist = int(ndf['Bus_perSqKm'][0])
inter_pop= int(ndf['Inter_perThouCap'][0])

print ("Class:",class_map(lr.predict([[bus_dist, bus_pop, inter_dist, inter_pop]])))
print("Approximately 40% of the variability observed in the Walk Scores is explained by the bus stop and intersection densities.")